staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilclock "k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// The 'timeout' query parameter in the request URL has an invalid duration specifier
+	invalidTimeoutInURL = "invalid timeout specified in the request URL"
+)
+
+// WithRequestDeadline determines the timeout duration applicable to the given request and sets a new context
+// with the appropriate deadline.
+// auditWrapper provides an http.Handler that audits a failed request.
+// longRunning returns true if he given request is a long running request.
+// requestTimeoutMaximum specifies the default request timeout value.
+func WithRequestDeadline(handler http.Handler, failWithAuditBuilder FailWithAuditBuilderFunc, longRunning request.LongRunningRequestCheck,
+	negotiatedSerializer runtime.NegotiatedSerializer, requestTimeoutMaximum time.Duration) http.Handler {
+	return withRequestDeadline(handler, failWithAuditBuilder, longRunning, negotiatedSerializer, requestTimeoutMaximum, utilclock.RealClock{})
+}
+
+func withRequestDeadline(handler http.Handler, failWithAuditBuilder FailWithAuditBuilderFunc, longRunning request.LongRunningRequestCheck,
+	negotiatedSerializer runtime.NegotiatedSerializer, requestTimeoutMaximum time.Duration, clock utilclock.PassiveClock) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+
+		requestInfo, ok := request.RequestInfoFrom(ctx)
+		if !ok {
+			handleError(w, req, http.StatusInternalServerError, fmt.Errorf("no RequestInfo found in context, handler chain must be wrong"))
+			return
+		}
+		if longRunning(req, requestInfo) {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		userSpecifiedTimeout, ok, err := parseTimeout(req)
+		if err != nil {
+			statusErr := apierrors.NewBadRequest(fmt.Sprintf("%s", err.Error()))
+
+			klog.Errorf("Error - %s: %#v", err.Error(), req.RequestURI)
+
+			failed := failedErrorHandler(negotiatedSerializer, statusErr)
+			failWithAudit := failWithAuditBuilder.BuildHandler(failed, func() string {
+				return statusErr.Error()
+			})
+
+			failWithAudit.ServeHTTP(w, req)
```

Comment:
use case A: user specifies a malformed timeout value in the request url,

before: the request is failed with `400` and it is not audited.
now:  the request is failed with `400` and it is audited.

@liggitt i have addressed your concern, now the failed request is audited.

Reply:
I'll defer review to @sttts as long as the issues with the original PR are fixed and tested